`terraform show -json` crash

This was originally reported in the terraform-provider-aws repository as [Issue 8754](https://github.com/terraform-providers/terraform-provider-aws/issues/8754)

I am copying @amh-mw's report from there.

<!---
Please note the following potential times when an issue might be in Terraform core:

* [Configuration Language](https://www.terraform.io/docs/configuration/index.html) or resource ordering issues
* [State](https://www.terraform.io/docs/state/index.html) and [State Backend](https://www.terraform.io/docs/backends/index.html) issues
* [Provisioner](https://www.terraform.io/docs/provisioners/index.html) issues
* [Registry](https://registry.terraform.io/) issues
* Spans resources across multiple providers

If you are running into one of these scenarios, we recommend opening an issue in the [Terraform core repository](https://github.com/hashicorp/terraform/) instead.
--->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform Version

Terraform v0.12.0
+ provider.aws v2.9.0

### Affected Resource(s)

* aws_security_group
* aws_vpc

### Terraform Configuration Files

Apologies for any typos; manual redactions.

```hcl
terraform {
  backend "remote" {
    organization = "foo"

    workspaces {
      prefix = "foo-"
    }
  }
}

locals {
  cidr_blocks = [
    "1.2.3.4/32",
  ]

  ipv6_cidr_blocks = [
    "12::34/128",
  ]
}

resource "aws_vpc" "foo" {
  assign_generated_ipv6_cidr_block = true
  cidr_block                       = "172.0.0.0/16"
  enable_dns_support               = true
  enable_dns_hostnames             = true
}

resource "aws_security_group" "foo" {
  name   = "foo"
  vpc_id = aws_vpc.foo.id

  ingress {
    from_port        = 22
    to_port          = 22
    protocol         = "tcp"
    cidr_blocks = local.cidr_blocks
    ipv6_cidr_blocks = local.ipv6_cidr_blocks
  }

  ingress {
    from_port        = 22
    to_port          = 22
    protocol         = "tcp"
    cidr_blocks      = [aws_vpc.foo.cidr_block]
    ipv6_cidr_blocks = [aws_vpc.foo.ipv6_cidr_block]
  }

  ingress {
    from_port   = 80
    to_port     = 80
    protocol    = "tcp"
    cidr_blocks = ["172.0.0.0/8"]
    # Note lack of ipv6_cidr_blocks here?
  }

  ingress {
    from_port       = 80
    to_port         = 80
    protocol        = "tcp"
    security_groups = [aws_security_group.bar.id]
    # Note lack of ipv6_cidr_blocks here?
  }
}
```

### Debug Output

<!---
Please provide a link to a GitHub Gist containing the complete debug output. Please do NOT paste the debug output in the issue; just paste a link to the Gist.

To obtain the debug output, see the [Terraform documentation on debugging](https://www.terraform.io/docs/internals/debugging.html).
--->
Could be provided on request. HCL above is a rather small excerpt of a larger whole.

### Panic Output

https://gist.github.com/amh-mw/d09d28a26ef43c2ac0843434023aca6d

```
...
Plan: 0 to add, 3 to change, 0 to destroy.
Failed running command "cd /terraform; envdir /env terraform show -json /terraform/terraform.tfplan" (exit 2)
Output: panic: inconsistent set element types (
cty.Object(map[string]cty.Type{"cidr_blocks":cty.List(cty.Bool), "description":cty.Bool, "from_port":cty.Bool, "ipv6_cidr_blocks":cty.List(cty.Bool), "prefix_list_ids":cty.List(cty.String), "protocol":cty.Bool, "security_groups":cty.Set(cty.String), "self":cty.Bool, "to_port":cty.Bool}) then
cty.Object(map[string]cty.Type{"cidr_blocks":cty.List(cty.Bool), "description":cty.Bool, "from_port":cty.Bool, "ipv6_cidr_blocks":cty.List(cty.String), "prefix_list_ids":cty.List(cty.String), "protocol":cty.Bool, "security_groups":cty.Set(cty.String), "self":cty.Bool, "to_port":cty.Bool}))
                                                                                                                                                 ^
LOOK HERE ---> ---> ---> ---> ---> ---> ---> ---> ---> ---> ---> ---> ---> ---> ---> ---> ---> ---> ---> ---> ---> ---> ---> ---> ---> ---> ---> |
```

### Expected Behavior

No crash.

### Actual Behavior

Crash.

### Steps to Reproduce

1. `terraform plan`

### Important Factoids

Just upgraded from 0.11.4 to 0.12.0 per https://www.terraform.io/upgrade-guides/0-12.html
Running against TFE/d377923
I deleted all egress/ingress stanzas and ran plan again, same crash.

### References

* None yet.

I'm still fighting with this issue. Check node_cname below, it's either a list of strings or an empty list.

I built commit ff9235c4642e39255dcf1f8bc7998c1458d658ad which should contain the fix.

panic: inconsistent list element types (
cty.Object(map[string]cty.Type{"address_type":cty.Bool, "default_gw":cty.Bool, "dns_flag":cty.Bool, "ilo_flag":cty.Bool, "mgmt_flag":cty.Bool, "node_cname":cty.List(cty.String), "node_domain":cty.Bool, "node_ip":cty.Bool, "node_name":cty.Bool, "subnet_mask":cty.Bool, "subnet_prefix_length":cty.Bool, "subnet_vlan":cty.Bool})
 then 
cty.Object(map[string]cty.Type{"address_type":cty.Bool, "default_gw":cty.Bool, "dns_flag":cty.Bool, "ilo_flag":cty.Bool, "mgmt_flag":cty.Bool, "node_cname":cty.List(cty.Bool), "node_domain":cty.Bool, "node_ip":cty.Bool, "node_name":cty.Bool, "subnet_mask":cty.Bool, "subnet_prefix_length":cty.Bool, "subnet_vlan":cty.Bool})
)

https://gist.github.com/marcsauter/4c8177522cd4520ac5ef2e4a957be50a

--- a/command/jsonplan/plan.go
+++ b/command/jsonplan/plan.go
@@ -215,10 +215,6 @@ func (p *plan) marshalResourceChanges(changes *plans.Changes, schemas *terraform
 						return cty.False, nil
 					}
 
-					if !val.Type().IsPrimitiveType() {
-						return val, nil // just pass through non-primitives; they already contain our transform results
-					}
-
 					if val.IsKnown() {
 						return cty.False, nil
 					}

--- a/command/jsonplan/plan.go
+++ b/command/jsonplan/plan.go
@@ -210,17 +210,7 @@ func (p *plan) marshalResourceChanges(changes *plans.Changes, schemas *terraform
 				if err != nil {
 					return err
 				}
-				afterUnknown, _ = cty.Transform(changeV.After, func(path cty.Path, val cty.Value) (cty.Value, error) {
-					if val.IsNull() {
-						return cty.False, nil
-					}
-
-					if val.IsKnown() {
-						return cty.False, nil
-					}
-
-					return cty.True, nil
-				})
+				afterUnknown = cty.EmptyObjectVal
 			} else {
 				filteredAfter := omitUnknowns(changeV.After)
 				if filteredAfter.IsNull() {

--- a/command/jsonplan/values_test.go
+++ b/command/jsonplan/values_test.go
@@ -173,7 +173,7 @@ func TestMarshalPlanResources(t *testing.T) {
 		Want   []resource
 		Err    bool
 	}{
-		"create with unkonwns": {
+		"create with unknowns": {
 			Action: plans.Create,
 			Before: cty.NullVal(cty.EmptyObject),
 			After: cty.ObjectVal(map[string]cty.Value{

--- a/command/show_test.go
+++ b/command/show_test.go
@@ -2,7 +2,6 @@ package command
 
 import (
 	"encoding/json"
-	"fmt"
 	"io/ioutil"
 	"os"
 	"path/filepath"
@@ -221,7 +220,6 @@ func TestShow_json_output(t *testing.T) {
 			json.Unmarshal([]byte(byteValue), &want)
 
 			if !cmp.Equal(got, want) {
-				fmt.Println(ui.OutputWriter.String())
 				t.Fatalf("wrong result:\n %v\n", cmp.Diff(got, want))
 			}
 

--- a/command/test-fixtures/show-json/basic-delete/output.json
+++ b/command/test-fixtures/show-json/basic-delete/output.json
@@ -48,10 +48,7 @@
                     "ami": "bar",
                     "id": "placeholder"
                 },
-                "after_unknown": {
-                    "ami": false,
-                    "id": false
-                }
+                "after_unknown": {}
             }
         },
         {
@@ -69,7 +66,7 @@
                     "id": "placeholder"
                 },
                 "after": null,
-                "after_unknown": false
+                "after_unknown": {}
             }
         }
     ],

--- a/command/test-fixtures/show-json/basic-update/output.json
+++ b/command/test-fixtures/show-json/basic-update/output.json
@@ -48,10 +48,7 @@
                     "ami": "bar",
                     "id": "placeholder"
                 },
-                "after_unknown": {
-                    "ami": false,
-                    "id": false
-                }
+                "after_unknown": {}
             }
         }
     ],

